abdiff/core/download_input_files.py

```diff
@@ -0,0 +1,85 @@
+import logging
+import subprocess
+
+import boto3
+from botocore.exceptions import ClientError
+from mypy_boto3_s3.client import S3Client
+
+from abdiff.config import Config
+
+logger = logging.getLogger(__name__)
+
+CONFIG = Config()
+
+
+def download_input_files(input_files: list[str]) -> None:
+    """Download extract files from S3 to a local MinIO server.
+
+    For each file download, two AWS CLI commands are run by subprocess.
+    The output from the first command is piped to the second command.
+    These commands are further explained below:
+
+        1. Copy the contents from the input file and direct to stdout.
+           ```
+           aws s3 cp <input_file> -
+           ```
+
+        2. Given the stdout from the previous command as input, copy the contents
+           to a similarly named file on the local MinIO server.
+           ```
+           aws s3 cp --endpoint-url <minio_s3_url> --profile minio - <input_file>
+           ```
+
+    Note: An S3 client connected to the local MinIO server will check whether the
+          file exists prior to any download.
+    """
+    s3_client = boto3.client(
+        "s3",
+        endpoint_url=CONFIG.minio_s3_url,
+        aws_access_key_id=CONFIG.minio_root_user,
+        aws_secret_access_key=CONFIG.minio_root_password,
+    )
+
+    for input_file in input_files:
```

Comment:
Can we add some progress logging here, something like:

```python
for i, input_file in enumerate(input_files):
    ...
    ...
```

And then perhaps skip the other logging messages, and add one shared one at the end?  Something like:

```python
logger.info(f"Input file {i+1} / {len(input_files)}: '{input_file}' available locally for transformation.")
```

For very large runs, I think this consistent output would be helpful.  I suppose we lose the grain of skipped vs downloaded, but the ouput would be consistent.  Thoughts?

Reply:
Hmm, I agree that in the case of large runs, it would be hard to parse through the varying logs to identify files that were already in the S3 bucket (and downloads were skipped for). 🤔 I'm good with the change proposed above if @ehanson8 agrees with the change as well!

**Outside the scope of this PR:** It feels like a CLI command that informs / warns / reminds user to clean up their storage in the case of very large runs, but it seems like that CLI command would only know which directory to check via the `MINIO_S3_LOCAL_STORAGE` environment variable (i.e., if an initial diff with large number of input files was run where MinIO storage is set to Folder A and a second diff was planned where MinIO storage is set to Folder B, the CLI command wouldn't raise any warnings when checking Folder B). I might've brought this up in a call with @ghukill , but a safeguard of some sort to remind users of previously downloaded data feels like it'd be helpful. 🤔

Reply:
> Hmm, I agree that in the case of large runs, it would be hard to parse through the varying logs to identify files that were already in the S3 bucket (and downloads were skipped for). 🤔 I'm good with the change proposed above if @ehanson8 agrees with the change as well!

I'm more concerned with counting to be honest.  I'm fully preparing to pass 1.6k input files to this, and I imagine the downloading may take.... 2-3 hours?  I think we need something like:

```
...
Input file 422 / 1630: 'alma-2024-10-01-full....xml' available locally for transformation.
Input file 423 / 1630: 'alma-2024-10-02-full....xml' available locally for transformation.
...
```

in the output to know how far along it is.  Of course, the _second_ time I run it, that list would sail by as those files are already present in Minio.  ABDiff is chatty, and I think that's okay; it'd just barf out 1.6k lines informing me that files are locally available for use.  But the first time, I think the incrementing progress is super helpful.

Which brings up your next comment:

> Outside the scope of this PR: It feels like a CLI command that informs / warns / reminds user to clean up their storage in the case of very large runs, but it seems like that CLI command would only know which directory to check via the MINIO_S3_LOCAL_STORAGE environment variable...

I would definitely lean into the idea this is "extras" functionality, and ABDiff has no responsibility to inform the user that they may have 50gb - 75gb in their minio server.  Open to pushback here, but data that big, it's hard _not_ to remember it's on disk!

Reply:
I agree with @ghukill 's proposed changes!

Reply:
Please see the latest commit for updates to improve logging. The logs shown below depict a `rundiff` command that involves 2 valid input files and 1 invalid input file that results in a `RuntimeError`. Let me know what both of you think of the logged messages: 

```text
2024-11-07 11:54:38,856 INFO abdiff.cli.main() line 44: Logger 'root' configured with level=DEBUG
2024-11-07 11:54:38,856 INFO abdiff.cli.main() line 45: Running process
2024-11-07 11:54:38,859 INFO abdiff.core.init_run.init_run() line 24: Run directory created: output/jobs/test-minio/runs/2024-11-07_16-54-38
2024-11-07 11:54:39,030 INFO abdiff.extras.minio.download_input_files.download_input_files() line 49: Input file: 1 / 3: 's3://timdex-extract-dev-222053980223/libguides/libguides-2024-06-03-full-extracted-records-to-index.xml' available locally for transformation.
2024-11-07 11:54:39,034 INFO abdiff.extras.minio.download_input_files.download_input_files() line 49: Input file: 2 / 3: 's3://timdex-extract-dev-222053980223/dspace/dspace-2024-03-06-daily-extracted-records-to-index.xml' available locally for transformation.
2024-11-07 11:54:41,359 INFO abdiff.extras.minio.download_input_files.download_input_files() line 55: Input file: 3 / 3: 's3://timdex-extract-dev-222053980223/libguides/libguides-2024-11-07-daily-extracted-records-to-index.xml' failed to download.
2024-11-07 11:54:41,359 INFO abdiff.extras.minio.download_input_files.download_input_files() line 59: Available input files: 2, missing input files: 1.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/Documents/repos/transmogrifier-ab-diff/abdiff/cli.py", line 175, in run_diff
    download_input_files(input_files_list)
  File "/Users/jcuerdo/Documents/repos/transmogrifier-ab-diff/abdiff/extras/minio/download_input_files.py", line 64, in download_input_files
    raise RuntimeError(  # noqa: TRY003
RuntimeError: 1 input file(s) failed to download.
```

This change also led @ghukill and I to push back the work to [support concurrent file downloads](https://mitlibraries.atlassian.net/browse/TIMX-386) to give us more time to think about how we can maintain effective logging (i.e., concurrent downloads may result in disordered 'counts' in log messages).